### DISCUSSION TITLE
Correct method for non-blocking socket acceptance

### DISCUSSION THREAD

**Discussion by dereknbartram** [micropython/micropython#10942] [RP2040 / Pico]
Hi,

What's the correct way to determine if sockets are pending?

```py
tcp_socket = socket.socket()
tcp_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
tcp_socket.bind(socket.getaddrinfo('0.0.0.0', 80)[0][-1])
tcp_socket.listen(99)

if ???CONNECTION IS AVAILABLE???:
   cl, addr = tcp_socket.accept()
```

This is essentially what I'm trying to achieve: if a connection is available, accept it and do stuff with it, otherwise do other stuff. The 'other stuff' requires as much processing time as possible, so I don't want to sit waiting for timeouts. In C# the TcpListener class has a 'Pending' property which indicates if 'accept' has anything to do, but I cannot find anything corresponding to that in micropython

The only option I have found is:
```py
tcp_socket.settimeout(SOCKET_TIMEOUT)
```
The accept call then terminates with _None_ if no socket pending, but if I set that sufficiently short, I'm finding when actually processing valid connections they then timeout - it's like the timeout applies to the complete life of the socket and not just 'acceptance'. I've tried setting tcp_socket.settimeout(0) and cl.settimeout(0.5) was similar issues.

**Comment by bixb922** [comment]
I believe this example could be of help

https://github.com/peterhinch/micropython-async/blob/master/v3/as_drivers/client_server/userver.py

**Reply by dereknbartram** [reply]
Thanks, that's doing essentially the same thing though; see line 34

**Reply by karfas** [reply]
Check out `select.poll()` at https://docs.micropython.org/en/latest/library/select.html.
This is what the async streams use internally.
